src/coreComponents/physicsSolvers/multiphysics/LagrangianContactSolver.cpp

```diff
@@ -1040,6 +1108,173 @@ real64 LagrangianContactSolver::CalculateResidualNorm( DomainPartition const & d
+                                                   m_solidSolver->getRigidBodyModes() );
+      }
+    }
+
+    // Preconditioner for the Schur complement: mechPrecond
+    std::unique_ptr< PreconditionerBase< LAInterface > > mechPrecond = LAInterface::createPreconditioner( mechParams, m_solidSolver->getRigidBodyModes() );
+    precond->setupBlock( 1,
+                         { { keys::TotalDisplacement, 0, 3 } },
+                         std::move( mechPrecond ) );
+
+    m_precond = std::move( precond );
+  }
+  else
+  {
+    //TODO: Revisit this part such that is coherent across physics solver
+    //m_precond = LAInterface::createPreconditioner( m_linearSolverParameters.get() );
+  }
+}
+
+void LagrangianContactSolver::ComputeRotationMatrices( DomainPartition & domain ) const
+{
+  GEOSX_MARK_FUNCTION;
+  MeshLevel & mesh = *domain.getMeshBody( 0 )->getMeshLevel( 0 );
+
+  FaceManager const & faceManager = *mesh.getFaceManager();
+  ElementRegionManager & elemManager = *mesh.getElemManager();
+
+  arrayView2d< real64 const > const & faceNormal = faceManager.faceNormal();
+
+  elemManager.forElementSubRegions< FaceElementSubRegion >( [&]( FaceElementSubRegion & subRegion )
+  {
+    if( subRegion.hasWrapper( m_tractionKey ) )
+    {
+      arrayView2d< localIndex const > const & elemsToFaces = subRegion.faceList();
+
+      arrayView3d< real64 > const &
+      rotationMatrix = subRegion.getReference< array3d< real64 > >( viewKeyStruct::rotationMatrixString );
+
+      forAll< serialPolicy >( subRegion.size(), [=]( localIndex const kfe )
+      {
+        array1d< real64 > Nbar( 3 );
+        Nbar[ 0 ] = faceNormal[elemsToFaces[kfe][0]][0] - faceNormal[elemsToFaces[kfe][1]][0];
+        Nbar[ 1 ] = faceNormal[elemsToFaces[kfe][0]][1] - faceNormal[elemsToFaces[kfe][1]][1];
+        Nbar[ 2 ] = faceNormal[elemsToFaces[kfe][0]][2] - faceNormal[elemsToFaces[kfe][1]][2];
+        LvArray::tensorOps::normalize< 3 >( Nbar );
+
+        computationalGeometry::RotationMatrix_3D( Nbar.toSliceConst(), rotationMatrix[kfe] );
+      } );
+    }
+  } );
+}
+
+void LagrangianContactSolver::ComputeFaceNodalArea( arrayView2d< real64 const, nodes::REFERENCE_POSITION_USD > const & nodePosition,
+                                                    ArrayOfArraysView< localIndex const > const & faceToNodeMap,
+                                                    localIndex const kf0,
+                                                    array1d< real64 > & nodalArea ) const
+{
+  // I've tried to access the finiteElement::dispatch3D with
+  // finiteElement::FiniteElementBase const &
+  // fe = fractureSubRegion->getReference< finiteElement::FiniteElementBase >( surfaceGenerator->getDiscretizationName() );
+  // but it's either empty (unknown discretization) or for 3D only (e.g., hexahedra)
+
+  localIndex const TriangularPermutation[3] = { 0, 1, 2 };
+  localIndex const QuadrilateralPermutation[4] = { 0, 1, 3, 2 };
+
```

Comment:
In the same order of the questions, answers are:
- you're right. See my answer above to a similar question;
- because I love symmetry ... so if I need a `QuadrilateralPermutation`, I would like to have a `TriangularPermutation`!
- I fully agree ... the missing correspondence creates a lot of small issues.

Reply:
> * because I love symmetry ... so if I need a `QuadrilateralPermutation`, I would like to have a `TriangularPermutation`!
> * I fully agree ... the missing correspondence creates a lot of small issues.

A consistent ordering will remove the need for the permutations. Correct?

Reply:
I hope the same node ordering logic is used....otherwise nothing should work.

Reply:
With reference to this sketch:
```
       2 o----------o 3
         |          | 
         |          | 
         |          | 
         |          | 
       0 o----------o 1
 ```
I have that
```
ArrayOfArraysView< localIndex const > const & faceToNodeMap = faceManager.nodeList().toViewConst();
```
gives me nodes in the following order: `0 1 3 2`, while `H1_QuadrilateralFace_Lagrange1_GaussLegendre2` requires them in the plot order, i.e., `0 1 2 3`.

Reply:
If you do keep those permutation arrays you should make them `constexpr`. Also it looks like you don't need the triangle one right?

Reply:
You're right. The triangular permutation can be removed.